File: node/src/main/java/dev.httpmarco.polocloud.node/commands/type/GroupArgument.java
package dev.httpmarco.polocloud.node.commands.type;

import dev.httpmarco.polocloud.api.Named;
import dev.httpmarco.polocloud.api.groups.ClusterGroup;
import dev.httpmarco.polocloud.node.commands.CommandArgument;
import dev.httpmarco.polocloud.api.groups.ClusterGroupProvider;
import dev.httpmarco.polocloud.node.commands.CommandContext;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Unmodifiable;

import java.util.List;
import java.util.Objects;

public final class GroupArgument extends CommandArgument<ClusterGroup> {

    private final ClusterGroupProvider groupService;

    public GroupArgument(String key, ClusterGroupProvider groupService) {
        super(key);
        this.groupService = groupService;
    }

    @Override
    public boolean predication(@NotNull String rawInput) {
        return groupService.exists(rawInput);
    }

    @Contract(pure = true)
    @Override
    public @NotNull String wrongReason() {
        return "The Argument " + key() + " is not a registered cluster group!";
    }

    @Override
    public @NotNull @Unmodifiable List<String> defaultArgs(CommandContext context) {
        return groupService.groups().stream().map(Named::name).toList();
    }

    @Contract("_ -> new")
    @Override
    public @NotNull ClusterGroup buildResult(String input) {
        return Objects.requireNonNull(groupService.find(input));
    }
}

File: agent/src/main/kotlin/dev/httpmarco/polocloud/agent/runtime/local/terminal/arguments/type/GroupArgument.kt
package dev.httpmarco.polocloud.agent.runtime.local.terminal.arguments.type

import dev.httpmarco.polocloud.agent.Agent
import dev.httpmarco.polocloud.agent.groups.AbstractGroup
import dev.httpmarco.polocloud.agent.runtime.local.terminal.arguments.TerminalArgument
import dev.httpmarco.polocloud.agent.runtime.local.terminal.arguments.InputContext

class GroupArgument(key: String = "group") : TerminalArgument<AbstractGroup>(key) {

    override fun buildResult(input: String, context: InputContext): AbstractGroup {
        return Agent.runtime.groupStorage().find(input)!!
    }

    override fun defaultArgs(context: InputContext): MutableList<String> {
        return Agent.runtime.groupStorage().findAll().map { it.name }.toMutableList()
    }

    override fun predication(rawInput: String): Boolean {
        return Agent.runtime.groupStorage().find(rawInput) != null
    }
}